Invitation error feedback

nickoferrall opened the pull request:
This is a PR for issue #4038.

### Updates

- Displaying a warning message with a warning icon if the user enters the email address of an existing teammate
- If the user has submitted the invitation and there's an error, display the error message with the error icon
- Gives the user feedback as to how many of the email addresses that they've entered are already existing teammates
- Fixed a bug in `parseEmailAddressList` where `emailAddresses.parseAddressList()` would return null if any email was invalid. This would cause the error message to flash as the user typed. 

### Test

- [ ] Click invite to team and enter the email address of an existing teammate. A warning message should display.
- [ ] Enter the email addresses of 2/3 existing teammates. The warning message should update.
- [ ] Once the warning message has displayed, type invalid emails before or after and the warning message shouldn't flash

nickoferrall commented on packages/client/utils/parseEmailAddressList.ts:
```diff
@@ -11,7 +12,20 @@ const parseEmailAddressList = (rawStr = '') => {
     .replace(/,$/g, '')
-  return emailAddresses.parseAddressList(commaDelimStr)
+
+  const commaDelimArr = commaDelimStr.split(', ')
```
True! I've removed the spaces and pushed the update. 

For the alternate solution, if we had a raw string of "validEmail1, invalidEmail, validEmail2" and it returned null, if we reduce the raw string by 1 character until it's no longer null, wouldn't we be left with only "validEmail1", i.e. we'd lose "validEmail2"?

mattkrick replied:
that's correct, i think it's OK if we say something like "invalid email(s) after validEmail1" because if they goof on the syntax then it's possible that goof is what made validEmail2 valid. e.g. `matt@parabol.co,b@ b@b.co`. we can't guarantee that they meant `b@b.co`, maybe they meant bb@b.co?

nickoferrall replied:
Fair point! I've added the `invalid email(s) after validEmail1` feedback while preventing the jumping. It's ready for a re-review when you are

ackernaut commented:
@nickoferrall do you need a review from me? I see Matt has made a few comments. Happy to do so.

nickoferrall commented:
@ackernaut I've added a new warning message colour, `WARNING_MAIN = '#FF9800'`, and added icons to the warning message. If you get the time, it'd be great to know if you feel it's a good fit for the design

ackernaut commented:
### Test cycle no. 1

- [ ] Click invite to team and enter the email address of an existing teammate. A warning message should display.
- [ ] Enter the email addresses of 2/3 existing teammates. The warning message should update.
- [ ] Once the warning message has displayed, type invalid emails before or after and the warning message shouldn't flash

ackernaut commented:
I pushed 2 small changes:
- IMO there’s not enough of a case to deviate from the brand palette for the warning color, so keeping it simple.
- I scrubbed the `rem` values. The forward-motion deprecation pattern we use is to clean things up as you see them, and not go scrubbing for the sake of scrubbing.

ackernaut commented:
Thanks for the efforts here, and sorry for the delay on the review 👋

nickoferrall commented:
Thanks for the review Terry - will stick with pixels over rem from now on 👍 

@mattkrick, when you get the chance, it'd be great to your eyes on the functionality of this one to make sure that there's no red flags

mattkrick commented on packages/client/components/AddTeamMemberModal.tsx:
```diff
@@ -98,34 +117,59 @@ const AddTeamMemberModal = (props: Props) => {
     const nextValue = e.target.value
     if (rawInvitees === nextValue) return
-    const parsedInvitees = parseEmailAddressList(nextValue)
+    const {parsedInvitees, invalidEmailExists} = parseEmailAddressList(nextValue) as any
```
it'd be good to do the typescript casting within the parseEmailAddressList instead of outside of it.
that way you can reuse the function anywhere & you still get type safety

mattkrick commented on packages/client/components/AddTeamMemberModal.tsx:
```diff
@@ -98,34 +117,59 @@ const AddTeamMemberModal = (props: Props) => {
+      lastValidEmail
+        ? onError(new Error(`Invalid email(s) after ${lastValidEmail}`))
+        : onError(new Error(`Invalid email(s)`))
+    } else if (!invalidEmailExists) {
```
can just be else

mattkrick commented on packages/client/utils/parseEmailAddressList.ts:
```diff
@@ -11,7 +19,38 @@ const parseEmailAddressList = (rawStr = '') => {
+  const filteredStr = isAddingNewEmail ? filterLastEmail(commaDelimStr) : commaDelimStr
+  if (emailAddresses.parseAddressList(filteredStr)) {
+    return {
+      parsedInvitees: emailAddresses.parseAddressList(filteredStr),
```
no need to parse twice. parse it & then perform the conditional

mattkrick commented on packages/client/utils/parseEmailAddressList.ts:
```diff
@@ -11,7 +19,38 @@ const parseEmailAddressList = (rawStr = '') => {
+  } else {
+    for (let i = filteredStr.length; i >= 0; i--) {
+      const slicedStr = filteredStr.slice(0, i)
+      if (emailAddresses.parseAddressList(slicedStr)) {
```
should only parse once. work performed on keystroke will feel laggy on mobile, so we wanna keep it as cheap as possible

mattkrick commented on packages/client/utils/parseEmailAddressList.ts:
```diff
@@ -11,7 +19,38 @@ const parseEmailAddressList = (rawStr = '') => {
+      }
+    }
+    return {
+      parsedInvitees: emailAddresses.parseAddressList(filteredStr),
```
parse once

mattkrick commented on packages/client/utils/parseEmailAddressList.ts:
```diff
@@ -1,7 +1,15 @@
+  return trimmedStr.slice(0, lastEmailIndex).replace(/,$/g, '')
+}
+
 const parseEmailAddressList = (rawStr = '') => {
```
define the return type here so we don't have to cast as `any` when we use it

mattkrick commented:
@nickoferrall reviewed! just a few minor things. nothing to block pushing to prod, but a couple things to clean it up a bit as a 2nd pass

nickoferrall commented:
Thanks @mattkrick! I'll add these changes when I create a PR for the bulk email bug
